ge0/parser.cpp

```diff
@@ -50,29 +50,55 @@ bool Ge0Parser::Parse(std::string const & url, Result & result)
 
 bool Ge0Parser::ParseAfterPrefix(std::string const & url, size_t from, Result & result)
 {
-  std::string remaining = url.substr(from);
-  if (remaining.find(',') != std::string::npos) {
-    // Optional slash after coordinates
-    size_t slashPos = remaining.find('/');
-    std::string coords = (slashPos == std::string::npos) ? remaining : remaining.substr(0, slashPos);
-
-    size_t commaPos = coords.find(',');
-    if (commaPos == std::string::npos)
-    return false;
-
-    std::string latStr = coords.substr(0, commaPos);
-    std::string lonStr = coords.substr(commaPos + 1);
-
-    double lat, lon;
-    if (!strings::to_double(latStr, lat) || !strings::to_double(lonStr, lon))
+  std::string_view remaining(url.data() + from, url.size() - from);
+  size_t posSep = remaining.find_first_of("?#");
+  if (posSep != std::string_view::npos)
+    remaining = remaining.substr(0, posSep);
+  if (remaining.empty())
     return false;
-
+  if (remaining.find(',') != std::string_view::npos)
+  {
+    std::string_view coords;
+    std::string_view name;
+    size_t slashPos = remaining.find('/');
+    if (slashPos == std::string_view::npos)
+    {
+      coords = remaining;
+    }
+    else
+    {
+      coords = remaining.substr(0, slashPos);
+      name = remaining.substr(slashPos + 1);
+      while (!name.empty() && name.front() == '/')
+        name.remove_prefix(1);
+      while (!name.empty() && name.back() == '/')
+        name.remove_suffix(1);
+    }
+    double lat, lon, zoom;
+    int pos = 0;
+    int count = sscanf(coords.data(), " %lf , %lf %n", &lat, &lon, &pos);
+    if (count == 2 && pos == (int)coords.size())
+    {
+      zoom = 18.0;
+    }
+    else
+    {
+      pos = 0;
+      count = sscanf(coords.data(), " %lf , %lf , %lf %n", &lat, &lon, &zoom, &pos);
+      if (count != 3 || pos != (int)coords.size())
+        return false;
+    }
+    if (lat < -90.0 || lat > 90.0 || lon < -180.0 || lon > 180.0)
+      return false;
+    if (!std::isfinite(zoom) || zoom < 1.0 || zoom > 22.0)
+      return false;
```

Comment:
if zoom is out of range then would be better to use default instead of failing completely?
or even better use the closest in-range value
e.g. if its 0.5 then fix to 1
if its 22 then fix to 20